Persistent storage for log entries and state (#48)
haraldng/omnipaxos: docs/src/sequencepaxos/storage.md

            self.log.truncate(from_idx as usize);
            self.append_entries(entries)
        }

        fn set_promise(&mut self, n_prom: Ballot) {
            self.n_prom = n_prom;
        }

        fn set_decided_idx(&mut self, ld: u64) {
            self.ld = ld;
        }

        fn get_decided_idx(&self) -> u64 {
            self.ld
        }

        fn set_accepted_round(&mut self, na: Ballot) {
            self.acc_round = na;
        }

        fn get_accepted_round(&self) -> Ballot {
            self.acc_round
        }

        fn get_entries(&self, from: u64, to: u64) -> Vec<T> {
            self.log
                .get(from as usize..to as usize)
                .unwrap_or(&[])
                .to_vec()
        }

        fn get_log_len(&self) -> u64 {
            self.log.len() as u64
        }

        fn get_suffix(&self, from: u64) -> Vec<T> {
            match self.log.get(from as usize..) {
                Some(s) => s.to_vec(),
                None => vec![],
            }
        }

        fn get_promise(&self) -> Ballot {
            self.n_prom
        }
        ...
    }
```

## PersistentStorage

> `PersistentStorage` is a persistent storage implementation that stores the replicated log and the state of OmniPaxos. The struct uses [Commitlog](https://crates.io/crates/commitlog) to store replicated logs, for storing the OmniPaxos state users can choose between [rocksDB](https://crates.io/crates/rocksdb) or [sled](https://crates.io/crates/sled). Users can configure the path to log entries and OmniPaxos state, and storage-related options through `PersistentStorageConfig`. The configuraiton struct features a `default()` constructor for generating default configuration, and the constructor `with()` that takes path and storage options as arguments. We recommend the user to generate the default configuration and then setting the desired fields.

Reviewer: ```suggestion
`PersistentStorage` is a persistent storage implementation that stores the replicated log and the state of OmniPaxos. The struct uses [Commitlog](https://crates.io/crates/commitlog) to store the replicated log, and the state is stored on [sled](https://crates.io/crates/sled) by default. The state can be changed to be stored on [RocksDB](https://crates.io/crates/rocksdb) instead of sled by using the feature`rocksdb`. Users can configure the path to log entries and OmniPaxos state, and storage-related options through `PersistentStorageConfig`. The configuration struct features a `default()` constructor for generating default configuration, and the constructor `with()` that takes path and storage options as arguments. 
```

};
use commitlog::LogOptions;
use sled::{Config};

// user-defined configuration
let my_path = "my_storage"
let my_log_opts = LogOptions::new(my_path);
let mut my_sled_opts = Config::new();
my_sled_opts.path(self, my_path);

// generate default configuration and set user-defined configuration
let mut my_config = PersistentStorageConfig::default();
my_config.set_path(my_path);
my_config.set_commitlog_options(my_logopts);
my_config.set_sled_options(my_sled_opts);
```
The same configuration can also be done with the constructor that takes arguments:
```rust,edition2018,no_run,noplaypen
let my_path = "another_storage"
let my_logopts = LogOptions::new(my_path);
let mut my_sled_opts = Config::new();
my_sled_opts.path(my_path);
my_sled_opts.new(true);

// create configuration with given arguments
let my_config = PersistentStorageConfig::with(my_path, my_logopts, my_sled_opts);
```
## Selecting storage for OmniPaxos state
`PersistentStorage` is configured to use `sled` by default, users can enable `rocksDB` as storage by adding `--features rocksdb` to their cargo command. The example below runs tests with rocksDB enabled:

```no_run,noplaypen
cargo test --features rocksdb

> ```

Reviewer: Show an example of how to import omnipaxos_storage with the rocksdb feature in `Cargo.toml` instead.

# Storage

> You are free to use any storage implementation with `SequencePaxos`. The only requirement is that it implements the `Storage` trait. OmniPaxos includes the package `omnipaxos_storage` which provides two types of storage implementation, depending on the users need for fast reads and writes or persistance: `MemoryStorage` and `PersistentStorage`

Reviewer: ```suggestion
You are free to use any storage implementation with `SequencePaxos`. The only requirement is that it implements the `Storage` trait. OmniPaxos includes the package `omnipaxos_storage` which provides two types of storage implementation that work out of the box: `MemoryStorage` and `PersistentStorage`.

If you do decide to implement your own storage, we recommend taking a look at `MemoryStorage` as a reference for implementing the functions required by `Storage`.
```

            self.log.append(&mut e);
            self.get_log_len()
        }

        fn append_on_prefix(&mut self, from_idx: u64, entries: Vec<T>) -> u64 {
            self.log.truncate(from_idx as usize);
            self.append_entries(entries)
        }

        fn set_promise(&mut self, n_prom: Ballot) {
            self.n_prom = n_prom;
        }

        fn set_decided_idx(&mut self, ld: u64) {
            self.ld = ld;
        }

        fn get_decided_idx(&self) -> u64 {
            self.ld
        }

        fn set_accepted_round(&mut self, na: Ballot) {
            self.acc_round = na;
        }

        fn get_accepted_round(&self) -> Ballot {
            self.acc_round
        }

        fn get_entries(&self, from: u64, to: u64) -> &[T] {
            self.log.get(from as usize..to as usize).unwrap_or(&[])
        }

        fn get_log_len(&self) -> u64 {
            self.log.len() as u64
        }

        fn get_suffix(&self, from: u64) -> &[T] {
            match self.log.get(from as usize..) {
                Some(s) => s,
                None => &[],
            }
        }

        fn get_promise(&self) -> Ballot {
            self.n_prom

>         }

Reviewer: Please update this according to the changes you have made. For example `get_entries` now returns a vector instead of slice.

        
        ...
    }
```

## PersistentStorage

> An persistent disk-based storage implementation, `PersistentStorage` makes use of the [Commitlog](https://crates.io/crates/commitlog) and [rocksDB](https://crates.io/crates/rocksdb) libraries to store logged entries and replica state. Users can configure the path to store log entries, replica state and storage related options through the `PersistentStorageConfig` struct. 

Reviewer: ```suggestion
`PersistentStorage` is a persistent storage implementation that uses [Commitlog](https://crates.io/crates/commitlog) to store the replicated log and [rocksDB](https://crates.io/crates/rocksdb) to store the state of OmniPaxos. Users can configure the path to store log entries, replica state, and storage related options through the `PersistentStorageConfig` struct. 
```

            let entry_bytes = bincode::serialize(&entry).expect("Failed to serialize");
            self.commitlog
                .append_msg(entry_bytes)
                .expect("Failed to append log entry")
                + 1
        }

        fn append_entries(&mut self, entries: Vec<T>) -> u64 {
            let serialized = entries
                .into_iter()
                .map(|entry| bincode::serialize(&entry).expect("Failed to serialize"));
            let offset = self
                .commitlog
                .append(&mut MessageBuf::from_iter(serialized))
                .expect("Falied to append entries");
            offset.first() + offset.len() as u64
        }

        fn append_on_prefix(&mut self, from_idx: u64, entries: Vec<T>) -> u64 {
            self.commitlog
                .truncate(from_idx)
                .expect("Failed to truncate log");
            self.append_entries(entries)
        }

        fn get_entries(&self, from: u64, to: u64) -> Vec<T> {
            let buffer = self
                .commitlog
                .read(from, ReadLimit::default())
                .expect("Failed to read from log");

>             let mut entries = Vec::<T>::with_capacity((to - from) as usize);
>             let mut iter = buffer.iter();
>             for _ in from..to {
>                 let msg = iter.next().unwrap();
>                 entries.push(bincode::deserialize(msg.payload()).expect("Failed to deserialize"));
>             }
>             entries
>         }
> 
>         fn get_log_len(&self) -> u64 {
>             self.commitlog.next_offset()
>         }
> 
>         fn get_suffix(&self, from: u64) -> Vec<T> {
>             self.get_entries(from, self.commitlog.next_offset())
>         }
> 
>         fn get_promise(&self) -> Ballot {
>             let value = self.rocksdb.get(NPROM).expect("Failed to retrive 'NPROM'");
>             match value {
>                 Some(prom_bytes) => {
>                     let b_store: BallotStorage =
>                         FromBytes::read_from(prom_bytes.as_slice()).expect("Failed to deserialize");
>                     Ballot::with(b_store.n, b_store.priority, b_store.pid)
>                 }
>                 None => Ballot::default(),
>             }
>         }
> 
>         fn set_promise(&mut self, n_prom: Ballot) {
>             let ballot_store = BallotStorage::with(n_prom);
>             let prom_bytes = AsBytes::as_bytes(&ballot_store);
>             self.rocksdb
>                 .put(NPROM, prom_bytes)
>                 .expect("Failed to set 'NPROM'")
>         }
> 
>         fn get_decided_idx(&self) -> u64 {
>             let value = self
>                 .rocksdb
>                 .get(DECIDE)
>                 .expect("Failed to retrive 'DECIDE'");
>             match value {
>                 Some(ld_bytes) => {
>                     FromBytes::read_from(ld_bytes.as_slice()).expect("Failed to deserialize")
>                 }
>                 None => 0,
>             }
>         }
> 
>         fn set_decided_idx(&mut self, ld: u64) {
>             let ld_bytes = AsBytes::as_bytes(&ld);
>             self.rocksdb
>                 .put(DECIDE, ld_bytes)
>                 .expect("Failed to set 'DECIDE'");
>         }
>             
>         ...
>     }
> ```

Reviewer: We don't need to show this to the user. We showed it for MemoryStorage just as a reference for them to understand how to implement the Storage trait. However, I would like you to show examples instead on how to use the constructor and set the different configurations for PersistentStorage.

```suggestion
```

# Storage

> You are free to use any storage implementation with `SequencePaxos`. The only requirement is that it implements the `Storage` trait. OmniPaxos includes the package `omnipaxos_storage` which provides two types of storage implementation that work out of the box: `MemoryStorage` **and** `PersistentStorage`.

Reviewer: ```suggestion
You are free to use any storage implementation with `SequencePaxos`. The only requirement is that it implements the `Storage` trait. OmniPaxos includes the package `omnipaxos_storage` which provides two types of storage implementation that work out of the box: `MemoryStorage` and `PersistentStorage`.
```

```

> ## PersistentStorage

Reviewer: Could you add a subsection here that shows how to do fail-recovery? I.e. you re-intiate the persistent storage and recreate a SequencePaxos instance and then call `seq_paxos.fail_recovery()`

let my_config = PersistentStorageConfig::with(my_path, my_logopts, my_sled_opts);
```
## Importing `omnipaxos_storage`

> To use the package, we need to add `omnipaxos_storage` to the dependencies in the cargo file. If a user wishes to configure the options for the commit log, then the user must add `commitlog` as a dependency. For configuring the state storage, either `rocksdb` or `sled` must also be added. Otherwise the default constructor for `PersistentStorageConfig` lets the user create default options for the commit log and the state storage without importing any dependencies.

Reviewer: It makes more sense to show how to import omnipaxos_storage before showing MemoryStorage and PersistentStorage. So I suggest doing this modification and move it to here: https://github.com/haraldng/omnipaxos/blob/c4d3a83d46f8e78dac4e5cfe93e3eb226307f85e/docs/src/sequencepaxos/storage.md?plain=1#L3

```suggestion
To use the provided storage implementations, we need to add `omnipaxos_storage` to the dependencies in the cargo file.
```
